apps/challenges/aws_utils.py

```diff
@@ -766,57 +767,110 @@ def restart_workers_signal_callback(sender, instance, field_name, **kwargs):
-                settings.HOSTNAME, challenge.id
-            )
-            challenge_manage_url = "https://{}/web/challenges/challenge-page/{}/manage".format(
-                settings.HOSTNAME, challenge.id
+    if count != 1:
+        logger.warning(
+            "Worker(s) for challenge {} couldn't restart! Error: {}".format(
+                challenge.id, failures[0]["message"]
             )
+        )
+    else:
 
-            if field_name == "test_annotation":
-                file_updated = "Test Annotation"
-            elif field_name == "evaluation_script":
-                file_updated = "Evaluation script"
+        challenge_url = "https://{}/web/challenges/challenge-page/{}".format(
+            settings.HOSTNAME, challenge.id
+        )
+        challenge_manage_url = "https://{}/web/challenges/challenge-page/{}/manage".format(
+            settings.HOSTNAME, challenge.id
+        )
 
-            template_data = {
-                "CHALLENGE_NAME": challenge.title,
-                "CHALLENGE_MANAGE_URL": challenge_manage_url,
-                "CHALLENGE_URL": challenge_url,
-                "FILE_UPDATED": file_updated,
-            }
+        if field_name == "test_annotation":
+            file_updated = "Test Annotation"
+        elif field_name == "evaluation_script":
+            file_updated = "Evaluation script"
+
+        template_data = {
+            "CHALLENGE_NAME": challenge.title,
+            "CHALLENGE_MANAGE_URL": challenge_manage_url,
+            "CHALLENGE_URL": challenge_url,
+            "FILE_UPDATED": file_updated,
+        }
 
-            if challenge.image:
-                template_data["CHALLENGE_IMAGE_URL"] = challenge.image.url
+        if challenge.image:
+            template_data["CHALLENGE_IMAGE_URL"] = challenge.image.url
 
-            template_id = settings.SENDGRID_SETTINGS.get("TEMPLATES").get(
-                "WORKER_RESTART_EMAIL"
+        emails = challenge.creator.get_all_challenge_host_email()
+
+        template_id = settings.SENDGRID_SETTINGS.get("TEMPLATES").get(
+            "WORKERS_RESTARTING_NOTIFIER"
+        )
+        for email in emails:
+            send_email(
+                sender=settings.CLOUDCV_TEAM_EMAIL,
+                recipient=email,
+                template_id=template_id,
+                template_data=template_data,
             )
 
-            emails = challenge.creator.get_all_challenge_host_email()
-            for email in emails:
-                send_email(
-                    sender=settings.CLOUDCV_TEAM_EMAIL,
-                    recipient=email,
-                    template_id=template_id,
-                    template_data=template_data,
+        ecs = get_boto3_client("ecs", aws_keys)
+        queue_name = challenge.queue
+        service_name = "{}_service".format(queue_name)
+
+        # Checking if the new workers are launched and old workers are stopped.
+        for i in range(100):
```

Comment:
this doesn't seem to be the right way to poll for worker status. Can you look for alternative methods

Reply:
There was another way to use a waiter, but it seems suitable only for checking start and stop workers. The way restart works is more complex since it first starts new workers (thereby doubling the number of workers for a brief period) and then scaling back to normal by deleting the old workers.

Some of the waiters use descibe_service internally to know how many tasks are running, so I did the same approach but I'm checking for the unique combination of parameters which signals us that the workers have fully restarted since the time we initiated it.